app/src/main/java/com/nextcloud/talk/controllers/ChatController.kt

```diff
@@ -1432,15 +1459,82 @@ class ChatController(args: Bundle) : BaseController(args), MessagesListAdapter
+                                    override fun onSubscribe(d: Disposable) {
+                                    }
+
+                                    override fun onNext(t: ChatOverallSingleMessage) {
+                                        if (t.ocs.meta.statusCode == 202) {
+                                            Toast.makeText(context, R.string.nc_delete_message_leaked_to_matterbridge,
+                                                    Toast.LENGTH_LONG).show()
+                                        }
+                                    }
+
+                                    override fun onError(e: Throwable) {
+                                        Log.e(TAG, "Something went wrong when trying to delete message with id " +
+                                                message?.id, e)
+                                    }
+
+                                    override fun onComplete() {
+                                    }
+                                })
+                        true
+                    }
                     else -> false
                 }
             }
             inflate(R.menu.chat_message_menu)
+            menu.findItem(R.id.action_copy_message).isVisible = !(message as ChatMessage).isDeleted
             menu.findItem(R.id.action_reply_to_message).isVisible = (message as ChatMessage).replyable
-            show()
+            menu.findItem(R.id.action_delete_message).isVisible = isShowMessageDeletionButton(message)
+            if(menu.hasVisibleItems()){
+                show()
+            }
+        }
+    }
+
+    private fun setMessageAsDeleted(message: IMessage?) {
+        val messageTemp = message as ChatMessage
+        messageTemp.isDeleted = true
+
+        messageTemp.isOneToOneConversation = currentConversation?.type == Conversation.ConversationType.ROOM_TYPE_ONE_TO_ONE_CALL
+        messageTemp.isLinkPreviewAllowed = isLinkPreviewAllowed
+        messageTemp.activeUser = conversationUser
+
+        adapter?.update(messageTemp)
+    }
+
+    private fun isShowMessageDeletionButton(message: ChatMessage): Boolean {
+        if (conversationUser == null) {
+            return false
         }
+
+        val isServerAbleToDelete = conversationUser.hasSpreedFeatureCapability("delete-messages")
+
+        val sixHoursInMillis = 6 * 3600 * 1000
+        val isNewerThanSixHours = message.createdAt?.after(Date(System.currentTimeMillis() - sixHoursInMillis)) == true
+
+        val isNormalMessage = message.systemMessageType == ChatMessage.SystemMessageType.DUMMY
+
+        val isNotYetDeleted = !message.isDeleted
+
+        val isUserAllowedByPrivileges = if (message.actorId == conversationUser.userId) {
+            true
+        } else {
+            currentConversation!!.isParticipantOwnerOrModerator
+        }
+
+        return isServerAbleToDelete && isNormalMessage && isNewerThanSixHours && isNotYetDeleted &&
```

Comment:
You could also return always false in previous cases as in L1515.
E.g. if (message.isDeleted) return false;

Reply:
yep.
java.net.HttpURLConnection.HTTP_ACCEPTED

Reply:
check